source/sliding_puzzle.py

```diff
@@ -126,21 +126,47 @@ def solved(self) -> bool:
                 i += 1
         return True
 
-    def _join(self, row: List[PuzzlePiece]) -> List[str]:
-        """Join together a row of puzzle pieces into a complete list of lines."""
-        res = []
-        for i in range(len(row[0].data)):
-            res.append('')
-            for piece in row:
-                res[i] += piece.data[i]
-
-        return res
-
     def draw(self) -> str:
         """Draw the puzzle in its current state."""
-        output = []
-        for row in self.rows:
-            output.extend(self._join(row))
+        nbr_separators = len(self.rows) - 1
+        piece_width = self.rows[0][0].width
+        puzzle_width = piece_width * len(self.rows[0]) + nbr_separators
+
+        def build_border(start: str, middle: str, end: str) -> str:
+            res = start + '─' * puzzle_width + end
+            border_count = 1
+            for i in range(piece_width, (puzzle_width - piece_width)):
+                if i % piece_width == 0:
+                    pos = i + border_count
+                    res = res[:pos] + middle + res[pos + 1:]
+                    border_count += 1
+            return res
+
+        def _join(row: List[List[str]]) -> List[str]:
+            """Join together a row of puzzle pieces into a complete list of lines"""
+            res = []
+            for i in range(len(row[0])):
+                res.append('│')
+                for piece in row:
+                    res[i] += piece[i]
+            return res
+
+        def bordered(lines: List[str]) -> List[str]:
+            """Put a border after every line"""
+            return [line + '│' for line in lines]
+
+        # build the top border
+        output = [build_border(start='┌', middle='┬', end='┐')]
+
+        # build content
+        for index, row in enumerate(self.rows):
+            row = [bordered(piece.data) for piece in row]
+            output.extend(_join(row))
+            if index < len(self.rows) - 1:
+                output.extend([build_border(start="├", middle="┼", end="┤")])
+
+        # add bottom border
+        output += [build_border(start='└', middle="┴", end='┘')]
 
         return '\n'.join(output)
```

Comment:
The more I think about this the more I question wether we should leave it up to the game menu..

That way all rendering would be at the same place. That said, this seems good enough and I think that we should merge it (after I've tried it locally).

Reply:
this doesn't directy print to the screen but returns what need to be print, I think It is in the right place, maybe we could just rename the function, it doesn't really draw